proposals/NNNN-async-stream-backpressure.md

```diff
@@ -0,0 +1,784 @@
+able to bridge backpressured systems. Furthermore, `Async[Throwing]Stream` is
+the best name. Therefore, this proposal decided to provide new interfaces to
+`Async[Throwing]Stream`.
+
+### Stick with the current `Continuation` and `yield` naming
+
+The proposal decided against sticking to the current names since the existing
+names caused confusion to them being used in multiple places. Continuation was
+both used by the `AsyncStream` but also by Swift Concurrency via
+`CheckedContinuation` and `UnsafeContinuation`. Similarly, yield was used by
+both `AsyncStream.Continuation.yield()`, `Task.yield()` and the `yield` keyword.
+Having different names for these different concepts makes it easier to explain
+their usage. The currently proposed `write` names were choosen to align with the
+future direction of adding an `AsyncWriter` protocol. `Source` is a common name
+in flow based systems such as Akka. Other names that were considered:
+
+- `enqueue`
+- `send`
+
+### Provide the `onTermination` callback to the factory method
+
+During development of the new APIs, I first tried to provide the `onTermination`
+callback in the `makeStream` method. However, that showed significant usability
+problems in scenarios where one wants to store the source in a type and
+reference `self` in the `onTermination` closure at the same time; hence, I kept
+the current pattern of setting the `onTermination` closure on the source.
+
+### Provide a `onConsumerCancellation` callback
+
+During the pitch phase it was raised that we should provide a
+`onConsumerCancellation` callback which gets invoked once the asynchronous
+stream notices that the consuming task got cancelled. This callback could be
+used to customize how cancellation is handled by the stream e.g. one could
+imagine writing a few more elements to the stream before finishing it. Right now
+the stream immediately returns `nil` or throws a `CancellationError` when it
+notices cancellation. This proposal decided to not provide this customization
+because it opens up the possiblity that asynchronous streams are not terminating
+when implemented incorrectly. Additionally, asynchronous sequences are not the
+only place where task cancellation leads to an immediate error being thrown i.e.
+`Task.sleep()` does the same. Hence, the value of the asynchronous not
+terminating immediately brings little value when the next call in the iterating
+task might throw. However, the implementation is flexible enough to add this in
+the future and we can just default it to the current behaviour.
+
+### Create a custom type for the `Result` of the `onProduceMore` callback
+
+The `onProducerMore` callback takes a `Result<Void, Error>` which is used to
+indicate if the producer should produce more or if the asynchronous stream
+finished. We could introduce a new type for this but the proposal decided
+against it since it effectively is a result type.  
+
```

Comment:
Do you want to incorporate any of the discussion on initializer versus factory method ergonomics mentioned at the end of the pitch thread?

Reply:
I added a section and linked to https://github.com/apple/swift-evolution/blob/main/proposals/0388-async-stream-factory.md which explains why the factory method approach is desirable